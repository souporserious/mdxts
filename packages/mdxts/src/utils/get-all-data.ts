import parseTitle from 'title'
import { dirname, join, sep } from 'node:path'
import { readPackageUpSync } from 'read-package-up'
import type { ExportedDeclarations, Project } from 'ts-morph'
import { Directory, SourceFile } from 'ts-morph'
import { getSymbolDescription, resolveExpression } from '@tsxmod/utils'
import matter from 'gray-matter'

import { filePathToPathname } from './file-path-to-pathname'
import { getExamplesFromSourceFile } from './get-examples'
import { getExportedSourceFiles } from './get-exported-source-files'
import { getExportedTypes } from './get-exported-types'
import { getGitMetadata } from './get-git-metadata'
import { getMainExportDeclaration } from './get-main-export-declaration'
import { getNameFromDeclaration } from './get-name-from-declaration'
import { getSourcePath } from './get-source-path'
import { getSharedDirectoryPath } from './get-shared-directory-path'
import { getPackageMetadata } from './get-package-metadata'

export type Pathname = string

export type ModuleImport = Promise<Record<string, any>>

export type AllModules = Record<Pathname, ModuleImport>

export type ModuleData<Type extends { frontMatter: Record<string, any> }> = {
  title: string
  label: string
  description?: string
  order: number
  depth: number
  mdxPath?: string
  tsPath?: string
  pathname: string
  url: string
  previous?: { label: string; pathname: string }
  next?: { label: string; pathname: string }
  sourcePath: string
  executionEnvironment?: 'server' | 'client' | 'isomorphic'
  isMainExport?: boolean
  exportedTypes: (Omit<
    ReturnType<typeof getExportedTypes>[number],
    'filePath'
  > & {
    pathname: string
    sourcePath: string
    isMainExport: boolean
  })[]
  examples: ReturnType<typeof getExamplesFromSourceFile>
} & ReturnType<typeof getGitMetadata> &
  ('frontMatter' extends keyof Type
    ? Type
    : { frontMatter: Record<string, any> })

export function getAllData<Type extends { frontMatter: Record<string, any> }>({
  allModules,
  globPattern,
  project,
  sourceDirectory = 'src',
  baseDirectory,
  basePathname = '',
}: {
  /** A map of all MDX modules keyed by their pathname. */
  allModules: AllModules

  /** The glob pattern used to calculate `allModules`. */
  globPattern: string

  /** The ts-morph project to use for parsing source files. */
  project: Project

  /** The source directory used to calculate package export paths. */
  sourceDirectory?: string

  /** The base directory to use when calculating source paths. */
  baseDirectory?: string

  /** The base path to use when calculating navigation paths. */
  basePathname?: string
}) {
  const typeScriptSourceFiles = /ts(x)?/.test(globPattern)
    ? project.addSourceFilesAtPaths(globPattern)
    : null
  const allPaths = [
    ...Object.keys(allModules),
    ...(typeScriptSourceFiles?.map((file) => file.getFilePath()) ?? []),
  ]

  if (allPaths.length === 0) {
    throw new Error(
      `mdxts: Could not find any files matching ${globPattern}. Please provide a valid file pattern.`
    )
  }

  const sharedDirectoryPath = getSharedDirectoryPath(...allPaths)
  const packageMetadata = getPackageMetadata(...allPaths)
  let entrySourceFiles = project.addSourceFilesAtPaths(
    packageMetadata?.exports
      ? /** If package.json exports found use that for calculating public paths. */
        Object.keys(packageMetadata.exports).map((key) =>
          join(
            packageMetadata.directory,
            sourceDirectory,
            key,
            'index.{js,jsx,ts,tsx}'
          )
        )
      : /** Otherwise default to a common root index file. */
        join(sharedDirectoryPath, 'index.{js,jsx,ts,tsx}')
  )

  /** If no root index files exist, assume the top-level directory files are all public exports. */
  if (
    typeScriptSourceFiles &&
    !packageMetadata?.exports &&
    entrySourceFiles.length === 0
  ) {
    entrySourceFiles = project.addSourceFilesAtPaths(
      join(sharedDirectoryPath, '*.{js,jsx,ts,tsx}')
    )
  }

  const exportedSourceFiles = getExportedSourceFiles(entrySourceFiles)
  const allPublicPaths = entrySourceFiles
    .concat(exportedSourceFiles)
    .map((sourceFile) => sourceFile.getFilePath() as string)
    .concat(Object.keys(allModules))
    .filter((path) => !path.includes('.examples.tsx'))
  const allData: Record<Pathname, ModuleData<Type>> = {}
  const allPublicDeclarations: WeakMap<SourceFile, ExportedDeclarations[]> =
    new WeakMap()

  allPublicPaths.forEach((path) => {
    const type =
      path.endsWith('.ts') || path.endsWith('.tsx')
        ? 'ts'
        : path.endsWith('.md') || path.endsWith('.mdx')
          ? 'md'
          : null
    const pathname = filePathToPathname(
      path,
      baseDirectory,
      basePathname,
      packageMetadata?.name
    )
    const previouseData = allData[pathname]
    const sourceFile = project.addSourceFileAtPath(path)
    const sourceFileTitle = getSourceFileTitle(sourceFile)
    const sourcePath = getSourcePath(path)
    const metadata = getMetadata(sourceFile)
    const depth = pathname.split(sep).length - 2
    let title =
      type === 'md'
        ? findFirstHeading(sourceFile.getText()) || sourceFileTitle
        : sourceFileTitle
    let label
    let description

    if (metadata?.title) {
      title = metadata.title
    }

    if (metadata?.label) {
      label = metadata.label
    } else {
      label = title
    }

    if (metadata?.description) {
      description = metadata.description
    }

    /** Handle TypeScript source files */
    if (type === 'ts') {
      const isIndex = sourceFile.getBaseNameWithoutExtension() === 'index'

      /** Cache all public declarations for later use when processing re-exported source files below */
      if (isIndex) {
        Array.from(sourceFile.getExportedDeclarations()).forEach(
          ([, declarations]) => {
            declarations.forEach((declaration) => {
              const publicDeclarations = allPublicDeclarations.get(
                declaration.getSourceFile()
              )

              if (publicDeclarations) {
                publicDeclarations.push(declaration)
              } else {
                allPublicDeclarations.set(declaration.getSourceFile(), [
                  declaration,
                ])
              }
            })
          }
        )
      }

      const mainExportDeclaration = getMainExportDeclaration(sourceFile)
      const mainExportDeclarationSymbol = mainExportDeclaration?.getSymbol()
      const mainExportDeclarationName = mainExportDeclarationSymbol?.getName()
      const exportedTypes = getExportedTypes(
        sourceFile,
        allPublicDeclarations.get(sourceFile)
      ).map(({ filePath, ...fileExport }) => {
        const isPublic = allPublicPaths.includes(filePath)
        const pathname = filePathToPathname(
          isPublic ? filePath : sourceFile.getFilePath(),
          baseDirectory,
          basePathname,
          packageMetadata?.name
        )
        return {
          ...fileExport,
          pathname,
          sourcePath: getSourcePath(filePath),
          isMainExport: mainExportDeclarationName
            ? mainExportDeclarationName === fileExport.name
            : false,
        }
      })
      const examples = getExamplesFromSourceFile(sourceFile, allModules)
      const isMainExport = packageMetadata?.name
        ? basePathname
          ? pathname === join(sep, basePathname, packageMetadata.name)
          : pathname === join(sep, packageMetadata.name)
        : false
      const importDeclarations = sourceFile.getImportDeclarations()
      const isServerOnly = importDeclarations.some((importDeclaration) => {
        const moduleSpecifier = importDeclaration.getModuleSpecifierValue()
        return moduleSpecifier === 'server-only'
      })
      const isClientOnly = importDeclarations.some((importDeclaration) => {
        const moduleSpecifier = importDeclaration.getModuleSpecifierValue()
        return moduleSpecifier === 'client-only'
      })
      const executionEnvironment = isServerOnly
        ? 'server'
        : isClientOnly
          ? 'client'
          : 'isomorphic'

      if (mainExportDeclaration) {
        const declarationName = getNameFromDeclaration(mainExportDeclaration)
        if (declarationName) {
          title = declarationName
        }
      }

      if (!description && mainExportDeclarationSymbol) {
        const symbolDescription = getSymbolDescription(
          mainExportDeclarationSymbol
        )
        if (symbolDescription) {
          description = symbolDescription
        }
      }

      allData[pathname] = {
        ...previouseData,
        tsPath: path,
        exportedTypes,
        examples,
        title,
        label,
        description,
        depth,
        executionEnvironment,
        isMainExport,
        pathname,
        sourcePath,
      }
    }

    /** Handle MDX content */
    if (type === 'md') {
      const { data } = matter(sourceFile.getText())

      allData[pathname] = {
        ...previouseData,
        mdxPath: path,
        exportedTypes: previouseData?.exportedTypes || [],
        examples: previouseData?.examples || [],
        description: previouseData?.description || description,
        frontMatter: data,
        title,
        label,
        depth,
        pathname,
        sourcePath,
      }
    }
  })

  // Add order, this must be done after all data has been collected and added to the project above
  const sharedDirectory = project.addDirectoryAtPath(sharedDirectoryPath)
  const sourceFilesSortOrder = getDirectorySourceFilesOrder(
    sharedDirectory,
    allPublicPaths
  )

  Object.values(allData).forEach((value) => {
    const sourcePath = (value.tsPath || value.mdxPath)!
    const isIndexOrReadme = /index|readme/i.test(sourcePath)

    if (value.isMainExport) {
      value.order = 0
    } else if (isIndexOrReadme) {
      const directoryPath = dirname(sourcePath)
      if (directoryPath in sourceFilesSortOrder) {
        value.order = sourceFilesSortOrder[directoryPath]
      }
    } else {
      if (sourcePath in sourceFilesSortOrder) {
        value.order = sourceFilesSortOrder[sourcePath]
      }
    }
  })

  const sortedAndFilteredData = Object.entries(allData)
    .sort((a, b) => {
      // Give the main export the highest priority
      if (a[1].isMainExport) {
        return -1
      }
      if (b[1].isMainExport) {
        return 1
      }

      // Sort by order if available
      if (a[1].order && b[1].order) {
        return a[1].order - b[1].order
      }
      if (a[1].order) {
        return -1
      }
      if (b[1].order) {
        return 1
      }

      // Fallback to alphabetical order
      return a[0].localeCompare(b[0])
    })
    .filter(([, data]) => data.mdxPath || data.exportedTypes.length > 0)

  // Add previous/next data to each module
  sortedAndFilteredData.forEach(([, data], index) => {
    const previousData = sortedAndFilteredData[index - 1]
    const nextData = sortedAndFilteredData[index + 1]
    if (previousData) {
      data.previous = {
        label: previousData[1].label,
        pathname: previousData[1].pathname,
      }
    }
    if (nextData) {
      data.next = {
        label: nextData[1].label,
        pathname: nextData[1].pathname,
      }
    }
  })

  // Calculate aggregate git metadata for MDX and TypeScript modules
  const parsedData = sortedAndFilteredData.map(([pathname, data]) => {
    let gitMetadata: ReturnType<typeof getGitMetadata> | null = null

    // Only calculate git metadata if it's requested
    const lazyGitMetadata = () => {
      if (gitMetadata === null) {
        gitMetadata = getGitMetadata(
          [data.tsPath, data.mdxPath].filter(Boolean) as string[]
        )
      }
      return gitMetadata
    }

    return [
      pathname,
      {
        ...data,
        get createdAt() {
          return lazyGitMetadata().createdAt
        },
        get updatedAt() {
          return lazyGitMetadata().updatedAt
        },
        get authors() {
          return lazyGitMetadata().authors
        },
        get url() {
          const siteUrl = process.env.MDXTS_SITE_URL
          if (!siteUrl) {
            throw new Error(
              '[mdxts] The `siteUrl` option in the `mdxts/next` plugin is required to generate the `url` field.'
            )
          }
          return `${siteUrl}${pathname}`
        },
      },
    ]
  })

  return Object.fromEntries(parsedData) as Record<Pathname, ModuleData<Type>>
}

/** Returns the title of a source file based on its filename. */
function getSourceFileTitle(sourceFile: SourceFile) {
  const baseName = sourceFile
    .getBaseNameWithoutExtension()
    .replace(/\d+\.?/g, '')
  return parseTitleFromBaseName(
    /(readme|index)$/i.test(baseName)
      ? sourceFile.getDirectory().getBaseName()
      : baseName
  )
}

/** Parses a title from a base name. */
function parseTitleFromBaseName(baseName: string) {
  // preserve PascalCase and camelCase names
  if (/^[A-Za-z][a-zA-Z0-9]*$/.test(baseName)) {
    return baseName
  }
  return parseTitle(baseName).replace(/-/g, ' ') // replace dashes with spaces
}

/** Returns the first h1 heading in a Markdown string. */
function findFirstHeading(sourceFileText: string) {
  const headingRegex = /(^|\n)#\s(.+)/
  const match = sourceFileText.match(headingRegex)
  if (match) {
    return match[2]
  }
  return null
}

/** Returns the metadata from a source file. */
function getMetadata(sourceFile: SourceFile) {
  const metadataExport = sourceFile.getVariableDeclaration('metadata')
  if (metadataExport) {
    if (sourceFile.getExtension() === '.mdx') {
      // since we're working with MDX and not a TypeScript file we need to trim the export to remove any trailing MDX that might be present
      const trailingCurlyBraceIndex = metadataExport.getText().lastIndexOf('}')

      if (trailingCurlyBraceIndex > -1) {
        const originalText = metadataExport.getText()
        metadataExport.replaceWithText(
          metadataExport.getText().slice(0, trailingCurlyBraceIndex + 1)
        )
        const metadata = resolveExpression(metadataExport.getInitializer()!)
        metadataExport.replaceWithText(originalText)
        return metadata as Record<string, any>
      }
    } else {
      const metadata = resolveExpression(metadataExport.getInitializer()!)
      return metadata as Record<string, any>
    }
  }
  return null
}

/** Returns a map of source file paths to their sort order. */
function getDirectorySourceFilesOrder(
  directory: Directory,
  allPublicPaths: string[]
): Record<string, number> {
  const orderMap: Record<string, number> = {}
  traverseDirectory(directory, '', orderMap, new Set(), allPublicPaths)
  return orderMap
}

/** Recursively traverses a directory, adding each file to the order map. */
function traverseDirectory(
  directory: Directory,
  prefix: string,
  orderMap: Record<string, number>,
  seenBaseNames: Set<string>,
  allPublicPaths: string[]
) {
  const isRoot = prefix === ''
  let index = 1

  if (!isRoot) {
    orderMap[directory.getPath()] = parseFloat(prefix.slice(0, -1)) // Remove trailing dot from prefix and convert to float
  }

  const directories = directory
    .getDirectories()
    .sort((a, b) => a.getBaseName().localeCompare(b.getBaseName()))
  const files = directory
    .getSourceFiles()
    .filter((file) => allPublicPaths.includes(file.getFilePath()))

  // Iterate through all files in the current directory
  for (const file of files) {
    // Extract the base part of the file name up to the first period e.g. `Button` from `Button.test.tsx`
    const baseName = file.getBaseName().split('.').at(0)
    if (baseName && !seenBaseNames.has(baseName)) {
      orderMap[file.getFilePath()] = parseFloat(`${prefix}${index}`)
      seenBaseNames.add(baseName)
      index++
    } else {
      orderMap[file.getFilePath()] = parseFloat(`${prefix}${index - 1}`)
    }
  }

  // Iterate through subdirectories
  for (const subdirectory of directories) {
    traverseDirectory(
      subdirectory,
      `${prefix}${index}.`,
      orderMap,
      new Set(),
      allPublicPaths
    )
    index++
  }
}
