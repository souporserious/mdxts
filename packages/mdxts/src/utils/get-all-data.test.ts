import { Project, InMemoryFileSystemHost } from 'ts-morph'
import { getAllData } from './get-all-data'

const workingDirectory = '/Users/username/Code/mdxts'

jest.mock('./get-git-metadata', () => ({
  getGitMetadata: jest.fn().mockReturnValue({
    authors: [],
    createdAt: undefined,
    updatedAt: undefined,
  }),
}))

describe('getAllData', () => {
  beforeAll(() => {
    process.env.MDXTS_GIT_SOURCE = 'https://github.com/souporserious/mdxts'
    process.env.MDXTS_GIT_BRANCH = 'main'
    process.env.MDXTS_SITE_URL = 'https://mdxts.dev'
  })

  beforeEach(() => {
    jest.spyOn(process, 'cwd').mockReturnValue(workingDirectory)
    jest.resetModules()
  })

  afterEach(() => {
    jest.restoreAllMocks()
  })

  it('should initialize correctly with basic input', () => {
    const project = new Project({ useInMemoryFileSystem: true })

    project.createSourceFile(
      'components/Button.tsx',
      `/** Used for any type of user action, including navigation. */\nexport function Button() {}`
    )

    project.createSourceFile(
      'components/Button.mdx',
      `# Button\n\nButtons allow for users to take actions in your application.`
    )

    const allData = getAllData({
      project,
      allModules: {
        '/components/Button.mdx': Promise.resolve({
          default: () => {},
        }),
      },
      globPattern: 'components/*.(ts|tsx)',
      baseDirectory: 'components',
    })

    expect(allData).toMatchSnapshot()
  })

  it('orders based on files and directories', () => {
    const project = new Project({ useInMemoryFileSystem: true })

    project.createSourceFile(
      'package/components/Button.tsx',
      `export function Button() {}`
    )

    project.createSourceFile(
      'package/components/Card.tsx',
      `export function Card() {}`
    )

    project.createSourceFile(
      'package/hooks/usePressable.ts',
      `export function usePressable() {}`
    )

    project.createSourceFile(
      'package/hooks/useFocus.ts',
      `export function useFocus() {}`
    )

    const allData = getAllData({
      project,
      allModules: {
        '/package/components/Button.tsx': Promise.resolve({
          default: () => {},
        }),
        '/package/components/Card.tsx': Promise.resolve({
          default: () => {},
        }),
        '/package/hooks/usePressable.ts': Promise.resolve({
          default: () => {},
        }),
        '/package/hooks/useFocus.ts': Promise.resolve({
          default: () => {},
        }),
      },
      globPattern: 'package/**/*.(ts|tsx)',
      baseDirectory: 'package',
    })

    expect(allData['/components/button'].order).toBe(1.1)
    expect(allData['/components/card'].order).toBe(1.2)
    expect(allData['/hooks/use-focus'].order).toBe(2.1)
    expect(allData['/hooks/use-pressable'].order).toBe(2.2)
  })

  it('parses order from file path', () => {
    const allData = getDocsData()

    expect(allData['/docs/getting-started'].order).toBe(1)
    expect(allData['/docs/routing'].order).toBe(2)
    expect(allData['/docs/examples/authoring'].order).toBe(3.1)
    expect(allData['/docs/examples/rendering'].order).toBe(3.2)
  })

  it('adds previous and next pathnames', () => {
    const allData = getDocsData()

    expect(allData['/docs/getting-started'].previous).toBeUndefined()
    expect(allData['/docs/getting-started'].next?.pathname).toBe(
      '/docs/routing'
    )
    expect(allData['/docs/routing'].previous?.pathname).toBe(
      '/docs/getting-started'
    )
    expect(allData['/docs/routing'].next?.pathname).toBe(
      '/docs/examples/authoring'
    )
    expect(allData['/docs/examples/authoring'].previous?.pathname).toBe(
      '/docs/routing'
    )
    expect(allData['/docs/examples/authoring'].next?.pathname).toBe(
      '/docs/examples/rendering'
    )
    expect(allData['/docs/examples/rendering'].previous?.pathname).toBe(
      '/docs/examples/authoring'
    )
    expect(allData['/docs/examples/rendering'].next).toBeUndefined()
  })

  it('includes only public files based on package.json exports', (done) => {
    jest.isolateModules(() => {
      jest.mock('read-package-up', () => ({
        readPackageUpSync: () => ({
          packageJson: {
            name: 'mdxts',
            exports: {
              './components': {
                types: './dist/components/index.d.ts',
                import: './dist/components/index.js',
                require: './dist/cjs/components/index.js',
              },
            },
          },
          path: `${workingDirectory}/package.json`,
        }),
      }))

      import('./get-all-data').then(({ getAllData }) => {
        const fileSystem = new InMemoryFileSystemHost()
        const files = [
          {
            path: 'components/Button.tsx',
            content: `export const Button = () => {};`,
          },
          {
            path: 'components/Button.examples.tsx',
            content: `import { Button } from './Button';\nexport const Basic = () => {};`,
          },
          {
            path: 'components/Card.tsx',
            content: `export const Card = () => {};`,
          },
          {
            path: 'components/PrivateComponent.tsx',
            content: `export const PrivateComponent = () => {};`,
          },
          {
            path: 'components/index.ts',
            content: `export { Button } from './Button'; export { Card } from './Card';`,
          },
          {
            path: 'hooks/usePressable.ts',
            content: `export const usePressable = () => {};`,
          },
          {
            path: 'hooks/useFocus.ts',
            content: `export const useFocus = () => {};`,
          },
          {
            path: 'hooks/index.ts',
            content: `export * from './usePressable';\nexport * from './useFocus';`,
          },
        ]

        files.forEach((file) => {
          fileSystem.writeFileSync(
            `${workingDirectory}/src/${file.path}`,
            file.content
          )
        })

        const project = new Project({ fileSystem })

        const allData = getAllData({
          project,
          allModules: {
            [`${workingDirectory}/src/components/Button.examples.tsx`]:
              Promise.resolve({
                default: () => {},
              }),
          },
          globPattern: `${workingDirectory}/src/**/*.{ts,tsx}`,
          baseDirectory: 'src',
        })

        expect(allData['/components/button']).toBeDefined()
        expect(allData['/components/card']).toBeDefined()
        expect(allData['/components/private-component']).toBeUndefined()

        done()
      })
    })
  })

  it('includes files when no index file is present', () => {
    const fileSystem = new InMemoryFileSystemHost()
    const files = [
      {
        path: 'components/Button.tsx',
        content: `/** Used for any type of user action, including navigation. */\nexport function Button() {}`,
      },
      {
        path: 'components/Button.examples.tsx',
        content: `import { Button } from './Button';\nexport const Basic = () => {};`,
      },
    ]

    files.forEach((file) => {
      fileSystem.writeFileSync(
        `${workingDirectory}/src/${file.path}`,
        file.content
      )
    })

    const project = new Project({ fileSystem })

    const allData = getAllData({
      project,
      allModules: {
        [`${workingDirectory}/src/components/Button.examples.tsx`]:
          Promise.resolve({ Basic: () => {} }),
      },
      globPattern: '*.tsx',
      baseDirectory: 'src',
    })

    expect(allData['/components/button']).toBeDefined()
  })

  it('includes only public declarations based on index file exports', () => {
    const fileSystem = new InMemoryFileSystemHost()
    const files = [
      {
        path: 'components/Button.tsx',
        content: `export const Button = () => {};\n\nexport const PrivateComponent = () => {};`,
      },
      {
        path: 'components/index.ts',
        content: `export { Button } from './Button';`,
      },
    ]

    files.forEach((file) => {
      fileSystem.writeFileSync(
        `${workingDirectory}/src/${file.path}`,
        file.content
      )
    })

    const project = new Project({ fileSystem })

    const allData = getAllData({
      project,
      allModules: {},
      globPattern: `${workingDirectory}/src/**/*.{ts,tsx}`,
      baseDirectory: `src`,
    })
    const exportedTypes = allData['/components/button'].exportedTypes

    expect(
      exportedTypes.find((type) => type.name === 'PrivateComponent')
    ).toBeUndefined()
  })
})

function getDocsData() {
  const project = new Project({ useInMemoryFileSystem: true })

  project.createSourceFile(
    'docs/01.getting-started.mdx',
    `# Getting Started\n\nStart here.`
  )

  project.createSourceFile(
    'docs/02.routing.mdx',
    `# Routing\n\nHelpers for routing.`
  )

  project.createSourceFile(
    'docs/03.examples/01.authoring.mdx',
    `# Authoring\n\nExamples can be written alongside source code.`
  )

  project.createSourceFile(
    'docs/03.examples/02.rendering.mdx',
    `# Rendering\n\nExamples can be rendered in the documentation using a bundler.`
  )

  return getAllData({
    project,
    allModules: {
      '/docs/01.getting-started.mdx': Promise.resolve({
        default: () => {},
      }),
      '/docs/02.routing.mdx': Promise.resolve({
        default: () => {},
      }),
      '/docs/03.examples/01.authoring.mdx': Promise.resolve({
        default: () => {},
      }),
      '/docs/03.examples/02.rendering.mdx': Promise.resolve({
        default: () => {},
      }),
    },
    globPattern: '**/*.mdx',
  })
}
