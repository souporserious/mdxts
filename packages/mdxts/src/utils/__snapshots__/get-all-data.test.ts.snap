// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllData should initialize correctly with basic input 1`] = `
{
  "/button": {
    "authors": [],
    "createdAt": undefined,
    "depth": 0,
    "description": undefined,
    "examples": [],
    "exportedTypes": [],
    "label": "Button",
    "mdxPath": "/components/Button.mdx",
    "order": 1,
    "pathname": "/button",
    "sourcePath": "",
    "title": "Button",
    "updatedAt": undefined,
    "url": "https://mdxts.dev/button",
  },
}
`;
